### DISCUSSION TITLE
Space-dependent diffusion with external function

### DISCUSSION THREAD

**Discussion by francescabasini** [zwicker-group/py-pde#229] [General]
Hello Dr. Zwicker,

thank you for this useful package!
I am trying to solve a Fokker-Planck equation in 1d where drift and diffusion aee space dependent. 
$\frac{d P(x, t)}{dt} = -\nabla (P(x, t) V(x)) + (D(x)) \nabla^2 P(x, t) $
I was able to define my own initial condition for the distribution (P(x, 0)) and the drift (V(x)). For these, I supplied specific data values for each point on the grid which I calculated externally to the package.  I report my code (working!) for reference. 

However, in the code below, D is given as a constant over space but I would like to supply my own values for it as well. 
My diffusion does not have an analytical expression for D(x), as I obtained it as a spline which is why I could not use something similar to example 2.15 from py-pde docs.

```
grid = CartesianGrid([[p2_range[0], p2_range[-1]]], [len(p2_range)], periodic=False)
state_p0 = pde.ScalarField(grid, data=proj_gaus_p, label="p0") # set ic field from numpy array

# solve the equation and store the trajectory
storage = MemoryStorage()

# Dynamics projected
D = 0.01
velocity = pde.VectorField(grid, data=Omega_proj_Proper)  # set velocity field from numpy array

eq = pde.PDE({'P': f'-divergence(P * V) + {D} * laplace(P)'}, consts={'V': velocity}, bc = [{"value":0},{"value":0}])
```

Thank you for considering my question

**Comment by david-zwicker** [comment]
You're almost there – simply use `D` (or maybe another name, like `diffusion` if sympy complains) as a variable in the expression and supply the (interpolated) field denoting the diffusivity as a `ScalarField` using the `consts` argument, analogously to the velocity field.

**Reply by francescabasini** [reply]
Everything's working now, many thanks for promptly clearing my doubt!

**Reply by francescabasini** [reply]
Hi again,

I was wondering one more thing: given the FPe above, thus without specifying my own pde class, is it possible to accelerate the computations using numba? If yes, how?

Many thanks again for this package

**Reply by david-zwicker** [reply]
The computations should already be accelerated by numba. Do you encounter issues with this?

**Reply by francescabasini** [reply]
Oh, fantastic then! 
The computations are fast indeed but I guess I had a bit of confusion as of when it is possible to use numba and I thought the base pde was not using it or needed extra settings for it, hence I was looking for further speed up. 
Thanks again for clearing my doubt
